src/scripts/client/modules/gradebook.js

```diff
@@ -1,424 +1,387 @@
-          );
-
-      } catch ( e ) {}
-      gradebooks = gradebooks || {};
-      return gradebooks;
+  }
+
+  // Behavior ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+  function main(sessionData, $el) {
+    $($el).on('showing', function renderView() {
+      show(sessionData, $el);
+    });
+  }
+
+  function loadCachedGradebooks() {
+    let gradebooks;
+    try {
+      gradebooks = JSON.parse(
+        localStorage.getItem('cachedGradeBookData')
+      );
+    } catch (e) {
+      console.warn('Failed to restore gradebook data from localStorage');
     }
-
-    function loadCachedSession() {
-      let session;
-      try {
-          session = JSON.parse(
-              localStorage.getItem( 'cachedGradeBookDataSession' )
-          );
-      } catch ( e ) {}
-      session = session || { path: null };
-      return session;
+    gradebooks = gradebooks || {};
+    return gradebooks;
+  }
+
+  function loadCachedSession() {
+    let session;
+    try {
+      session = JSON.parse(
+        localStorage.getItem('cachedGradeBookDataSession')
+      );
+    } catch (e) {
+      console.warn('Failed to restore gradebook session data from localStorage');
     }
+    session = session || { path: null };
+    return session;
+  }
+
+  /**
+  Given any newline uri make an initial request, detect if there are more pages,
+  if they are re curse and build a complete collection.
+  **/
+  function recurseOverCollection(uri, collection, page) {
+    return new Promise(function collectCollections(res) {
+      const settings = {
+        async: true,
+        crossDomain: true,
+        url: `${uri}&page=${page}`,
+        method: 'GET'
+      };
 
-    function show( sessionData, $el ) {
-        if ( uiBuilt ) {
-            return;
-        }
-
-        let cachedSession = loadCachedSession();
-        console.log(sessionData, cachedSession);
-        if ( sessionData.path === null && cachedSession.path === null ) {
-            // Provide feedback to the user that we can't handle this.
-            $el.text( 'Have you visited a path recently, we don\'t know what gradebook you want to see, go to a path you own and try again?!' );
-            return;
-        } else if ( sessionData.path === null && cachedSession.path !== null  ) {
-          sessionData = cachedSession;
-        } else {
-          localStorage.setItem( 'cachedGradeBookDataSession', JSON.stringify(sessionData) );
-        }
-
-        let gradebooks = loadCachedGradebooks();
-        let gradebook_data = gradebooks[ sessionData.path.id ];
-
-        resetUI( sessionData, $el );
-
-        if ( gradebook_data ) {
-            try {
-                buildUI(
-                    sessionData,
-                    $el,
-                    gradebook_data
-                );
-            } catch ( e ) {
-                localStorage.removeItem( 'cachedGradeBookData' );
-            }
+      $.get(settings).done(function handleCollectionRequest(response) {
```

Comment:
Do we need a `.fail()` for this ajax call and then `reject` the Promise? Or do we never want to reject the promise? In any case, wouldn't we want to handle ajax failures?